### ISSUE TITLE
CLBlast: byte offset / element count confusion

### ISSUE THREAD

**Issue by shibe2** [ggerganov/llama.cpp#3307]
# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [YES] I am running the latest code. bc9d3e3971e5607a10ff4c24e39568ce1ac87271
- [YES] I carefully followed the [README.md](https://github.com/ggerganov/llama.cpp/blob/master/README.md).
- [YES] I [searched using keywords relevant to my issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/filtering-and-searching-issues-and-pull-requests) to make sure that I am creating a new issue that is not already open (or closed).
- [YES] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new bug or useful enhancement to share.

# Expected Behavior

Correct uploading of contiguous 3D tensor data to GPU.

# Current Behavior

`ggml_cl_h2d_tensor_2d` uses `offset` argument as byte offset in a call to `clEnqueueWriteBuffer`. `ggml_cl_transform_tensor` passes element count as `offset` to `ggml_cl_h2d_tensor_2d`. This corresponds to byte offset only if element size is exactly 1.

Also, I don't understand why `ggml_cl_mul_f32` passes non-zero offset to `ggml_cl_h2d_tensor_2d`.

# Environment and Context

AMD GPU
Linux

# Steps to Reproduce

1. Pass 3D tensor with contiguous `GGML_TYPE_F16` or `GGML_TYPE_F32` data to `ggml_cl_transform_tensor`.
2. Read data back from GPU memory or perform `ggml_cl_mul_mat` on that tensor.
3. Observe incorrect data or result.

# Ping

@0cc4m
@JohannesGaessler
@SlyEcho

**Comment by shibe2** [comment]
I think, it was was wrong from the beginning (since 2e6cd4b02549e343bef3768e6b946f999c82e823). With 2D tensors, offset is always 0, and the bug does not manifest itself. With 3D tensors, 2D slices partially overwrite previous slices.

**Comment by shibe2** [comment]
I fixed this specific bug by passing byte offset from `ggml_cl_transform_tensor` to `ggml_cl_h2d_tensor_2d`. I also fixed 2 related issues, so uploading works properly in all cases. Tested by reading data back from VRAM.

shibe2@8cd5b0aba63c34987757f8d3ef3afa27ee6e38b9

However, this is not usable on its own, because that data is addressed incorrectly during computation. I don't know if this fix is worth merging separately. Currently I have only a partial fix for the computation part.
